File: apps/admin-dashboard/lib/api/api-meta.ts
export type ApiMeta = {
  readonly timestamp?: string;
  readonly path?: string;
  readonly method?: string;
  readonly requestId?: string;
  readonly pagination?: Record<string, unknown>;
};

export function parseApiMeta(value: unknown): ApiMeta {
  if (!isObject(value)) {
    return {};
  }

  return {
    timestamp: asOptionalString(value.timestamp),
    path: asOptionalString(value.path),
    method: asOptionalString(value.method),
    requestId: asOptionalString(value.requestId),
    pagination: isObject(value.pagination)
      ? value.pagination
      : undefined,
  };
}

function asOptionalString(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: apps/mobile/lib/core/networking/api_meta.dart
class ApiMeta {
  const ApiMeta({
    this.timestamp,
    this.path,
    this.method,
    this.requestId,
    this.pagination,
  });

  factory ApiMeta.fromJson(Map<String, dynamic>? json) {
    if (json == null) {
      return const ApiMeta();
    }

    return ApiMeta(
      timestamp: json['timestamp'] as String?,
      path: json['path'] as String?,
      method: json['method'] as String?,
      requestId: json['requestId'] as String?,
      pagination: json['pagination'] is Map<String, dynamic>
          ? Map<String, dynamic>.unmodifiable(
              json['pagination'] as Map<String, dynamic>,
            )
          : null,
    );
  }

  final String? timestamp;
  final String? path;
  final String? method;
  final String? requestId;
  final Map<String, dynamic>? pagination;
}
